plutus-core/plutus-ir/src/PlutusIR/Transform/DeadCode.hs

```diff
@@ -25,54 +25,81 @@ import qualified Algebra.Graph                  as G
+removeDeadBindings ::
+    ( PLC.HasUnique name PLC.TermUnique
+    , PLC.HasUnique tyname PLC.TypeUnique
+    , PLC.ToBuiltinMeaning uni fun
+    , PLC.MonadQuote m
+    )
+    => TruncateTypes
+    -> Term tyname name uni fun a
     -> m (Term tyname name uni fun a)
-removeDeadBindings t = do
+removeDeadBindings tt t = do
     tRen <- PLC.rename t
-    runReaderT (transformMOf termSubterms processTerm tRen) (calculateLiveness tRen)
+    runReaderT (transformMOf termSubterms processTerm tRen)
+               (Ctx (calculateLiveness tRen) tt)
 
 type Liveness = Set.Set Deps.Node
+data Ctx = Ctx Liveness TruncateTypes
 
-calculateLiveness
-    :: (PLC.HasUnique name PLC.TermUnique, PLC.HasUnique tyname PLC.TypeUnique,
-       PLC.ToBuiltinMeaning uni fun)
+calculateLiveness ::
+    ( PLC.HasUnique name PLC.TermUnique
+    , PLC.HasUnique tyname PLC.TypeUnique
+    , PLC.ToBuiltinMeaning uni fun
+    )
     => Term tyname name uni fun a
     -> Liveness
 calculateLiveness t =
-    let
-        depGraph :: G.Graph Deps.Node
+    let depGraph :: G.Graph Deps.Node
         depGraph = fst $ Deps.runTermDeps t
     in Set.fromList $ T.reachable Deps.Root depGraph
 
-live :: (MonadReader Liveness m, PLC.HasUnique n unique) => n -> m Bool
-live n =
-    let
-        u = coerce $ n ^. PLC.unique
-    in asks $ Set.member (Deps.Variable u)
+live :: (MonadReader Ctx m, PLC.HasUnique n unique) => n -> m Bool
+live n = do
+    Ctx liveness _ <- ask
+    let u = coerce $ n ^. PLC.unique
+    pure $ Set.member (Deps.Variable u) liveness
 
-liveBinding
-    :: (MonadReader Liveness m, PLC.HasUnique name PLC.TermUnique, PLC.HasUnique tyname PLC.TypeUnique)
+-- TODO: Refactor drunk code
+liveBinding ::
+    ( MonadReader Ctx m
+    , PLC.HasUnique name PLC.TermUnique
+    , PLC.HasUnique tyname PLC.TypeUnique
+    )
     => Binding tyname name uni fun a
-    -> m Bool
-liveBinding =
-    let
-        -- TODO: HasUnique instances for VarDecl and TyVarDecl?
+    -> m (Maybe (Binding tyname name uni fun a))
+liveBinding b = do
+    Ctx _ truncateTypes <- ask
+    let -- TODO: HasUnique instances for VarDecl and TyVarDecl?
         liveVarDecl (VarDecl _ n _) = live n
         liveTyVarDecl (TyVarDecl _ n _) = live n
-    in \case
-        TermBind _ _ d _ -> liveVarDecl d
-        TypeBind _ d _ -> liveTyVarDecl d
-        DatatypeBind _ (Datatype _ d _ destr constrs) -> or <$> (sequence $ [liveTyVarDecl d,  live destr] ++ fmap liveVarDecl constrs)
+        f True = Just b
+        f _    = Nothing
+    case b of
+        TermBind _ _ d _ -> f <$> liveVarDecl d
+        TypeBind _ d _ -> f <$> liveTyVarDecl d
+        DatatypeBind a (Datatype a' d t destr constrs) ->
+            if truncateTypes -- Allow the removal of unused constructors
+            then liveTyVarDecl d >>= \case
+                True -> filterM liveVarDecl constrs >>= \constrs' ->
+                    -- #4147: An empty `constrs'` here basically means a type that is only used at the type level.
```

Comment:
> Well IMO that's a bug in your dependency analysis ;) In this program I claim that only the destructor should be live:
> ```
> let datatype Maybe = ...
> in \(x :: Maybe Integer) -> matchMaybe x (\y -> y) 0
> ```

During the last session, I realized that if there are no constructors in the program to create a single `Maybe` then that `matchMaybe` should also be dead and removed! There is nothing to match on and both its branches have to be removed.

Reply:
It's still not *dead* in the sense of not being mentioned anywhere. It's unclear how it could be implemented (I don't actually know what the compiler will do at the moment if you feed it a datatype with no constructors), but let's not get too hung up on that for now!

Reply:
True, there should be more practical and exciting approaches to try first!